### DISCUSSION TITLE
Inputs() are not discovered if part of a common directive after upgrade from Angular 13 to Angular 18

### DISCUSSION THREAD

**Discussion by Ghevi** [gund/ng-dynamic-component#518] [Q&A] [answered]
Hi, I make this a discussion because I'm not sure if it's a bug of the library or an intended change.
I have components that share some properties in this way, `MySharedClass` is not the only one that is extended:

````
export interface MyComponent extends MySharedClass, /*...other classes*/ { }

@Component({
...
})
export class MyComponent  {
  @Input() data: Data;
}

@Directive()
export class MySharedClass {
  @Input() sharedData: SharedData;
}
````
It was working fine in Angular 13 but in later versions it produces this error:
`NG0303: Can't set value of the 'sharedData' input on the 'MyComponent' component. Make sure that the 'sharedData' property is annotated with @Input() or a mapped @Input('sharedData') exists.`

If i put `@Input() sharedData: SharedData;` directly in the component, then it works again.
Unfortunately I was not able to test it immediately in Angular 14. I can try making a repo if it can help.

**Comment by gund** [comment]
This error is coming directly from Angular as we switched to a new API provided in v14.
So you need to make sure that all your inputs are properly annotated to be able to set them dynamically using this library.

**Reply by Ghevi** [reply]
Thank you for the response!
So just to be clear, by "properly annotated" do you mean they cannot be shared with these common traits/directives but should be specify in the component directly?
Because I also tried with passing a name to the Input() but it still give the same error.

**Reply by gund** [reply]
Well from your example it's not quite clear how exactly are you sharing the inputs, is it just by defined interface that your components implement or do you have angular directives that your components extend, in which case you won't be able to extend multiple directives as you can only extend a single parent but you can still do multilevel inheritance I guess which is less useful as it's no longer a trait/mixin pattern.
My point is - if you are trying to reuse inputs, then your components MUST inherit directives which declare those inputs, otherwise you would get those errors about unknown inputs.
There is also a chance that Angular changed something in latest versions which breaks the inheritance of the inputs, which I doubt but it's still a possibility, so if you are doing inheritance correctly and it still does not work - I suggest you to create a simple online repro so we can investigate it.
And a final thought, if you like you can completely bypass this error by setting the schema of your component where you are rendering dynamic components to [`NO_ERRORS_SCHEMA`](https://angular.dev/api/core/NO_ERRORS_SCHEMA) - it will simply not warn you about any inputs that you set in that component. This may be useful until you find a proper workaround.

**Reply by gund** [reply]
I did a [small repro](https://stackblitz.com/edit/stackblitz-starters-aabm9z?file=src%2Fmain.ts) with shared inputs from multiple directives via multilevel inheritance and it's working just fine.
So you either using just interfaces for inputs, which would obviously not work, or you only inherited some directives but not all.

**Reply by Ghevi** [reply]
Thank you for your time. I checked the repo and I get what you mean.

Unfortunately, I have this project which to me seems a bit complicated. So I can share the code where your component is used and how the directive's Inputs are shared, but I fully understand if you might now want to spend extra time diving into this specific code. 
This is still with Angular 13 and the error i mentioned do not appear, maybe I can just ignore them as you said with `NO_ERRORS_SCHEMA`. 
Of course thank you for any extra help you'll provide.

https://github.com/Ghevi/evt-viewer-angular/blob/master/src/app/components/paragraph/paragraph.component.ts
https://github.com/Ghevi/evt-viewer-angular/blob/master/src/app/components/components-mixins.ts
https://github.com/Ghevi/evt-viewer-angular/blob/master/src/app/components/content-viewer/content-viewer.component.html

**Reply by gund** [reply, marked as answer]
Well as I already explained, you cannot inherit inputs or anything really by extending interfaces with the same name and think that Angular will magically pickup those inputs somehow like you are doing [here](https://github.com/Ghevi/evt-viewer-angular/blob/f479cc078bbf04c383af8ea77b31b209da820825/src/app/components/paragraph/paragraph.component.ts#L7).
Interfaces have nothing to do with classes even when have the same name, you are just adding extra type information but nothing in the runtime is affected and Angular also does not care what you do with interfaces.
Like I already said above, your component class MUST extend another component/directive class not interfaces to inherit it's inputs and all the rest of the behaviors. You may refer to the example I provided to see how it looks like but the way you've built your "inheritance" is incorrect since you are just manipulating pure types not actual runtime classes.

**Reply by Ghevi** [reply]
I see, thank you for the extra clarification
